### ISSUE TITLE
# Ideas / thoughts about how to implement args.

### ISSUE THREAD

**Issue by Plommonsorbet** [pier-cli/pier#37]
The args feature has been mocked up for a while and I've had it in the back of my mind how it would be good to implement it, initially I was thinking you could just do some templating but I pretty soon realized that it's a pretty big security flaw as you could then literally inject code before you run it.

Right now I'm leaning towards the best option would be to implement arguments/options as env variables, perhaps via a key/value cli arg with the ability to set defaults for a tag/script in the configt.

I imagine the cli option for setting environment variables to something like -E/--env it could look something like this perhaps:
pier run build_something -E RUST_BACKTRACE:1 -E DB_HOST:127.0.0.1

**Comment by BenSchZA** [comment]
Definitely! This has been on the back-burner for a while.

I think for familiarity we can use Docker as an example:

```-e, --env list                       Set environment variables
      --env-file list                  Read in a file of environment variables
```

```docker run --env VAR1=value1 --env VAR2=value2```
```docker run --env VAR1 --env VAR2```

Re. templating - are you referring to loading the variables directly from the host environment? I think the above Docker example is an acceptable risk, and it's worth the convenience for users.

Being able to set defaults would be useful, maybe we can break that into a separate ticket.

**Comment by Plommonsorbet** [comment]
> I think for familiarity we can use Docker as an example:

That's a great idea, I think in general to try and keep the commandline flags inline with other common cli programs is a big win.

> Re. templating - are you referring to loading the variables directly from the host environment? I think the above Docker example is an acceptable risk, and it's worth the convenience for users.

What I mean was to actually insert the argument into the script before it is run not using environment variables.

**Below you have a few examples what I meant by templating and my proposed method:**

#### Upgrading and sending a custom notification.
```
[scripts.test-command]
alias = 'test-command'
command = '''
sudo apt upgrade
send-notify {{}}
'''
```
**How i imagine this being abused:**
`pier run test-command -a '" && sudo ... do something malicious... &&"'`

#### Another example of how this could be abused.

```
[scripts.test-command]
alias = 'test-command'
command = '''
sudo apt upgrade
send-notify "{{}}"
'''
```
**How I can imagine this being abused: **
`pier run test-command -a '$( sudo ... do something malicious... )'`

If you take my example above you are able to actually insert new syntax into the script being run. which in my opinion would make it very dangerous to have _any_ kinda sudo in your scripts. 

#### This is how I imagine a good way to go about implementing this:

##### Setting the level, summary and body of a notification
```
[scripts.test-command]
alias = 'test-command'
command = '''
sudo apt upgrade

send-notify -u $LEVEL "$1" "$2"
'''
```
`pier run  -e LEVEL=low -a "Notification summary" "Notification body"`

##### Example running a script and detatching it from itself.
```
[script.spawn-and-detach]
alias = 'spawn-and-detach'
command = '''
nohup $@ > /dev/null 2>&1&
'''
```
`pier run -a firefox`

We can still do things like run a binary, however it would mean that at least you're limited by the safeties of the runtime you are running. It would also be much more consisten with how the shell works in my opinion

> Being able to set defaults would be useful, maybe we can break that into a separate ticket.

That's fine by me, I think there's other options that could also be a great way to make tag specific options. Something from the top of my head is having the default_interpreter option from #36.

**Comment by Plommonsorbet** [comment]
Something that came to mind for how to define positional arguments, instead of defining them via ```-a xyz -a foo -a bar``` perhaps it would be possible to have the same syntax cargo uses? e.g.
```cargo run -- <arg1> --my-flag1```

**Comment by BenSchZA** [comment]
I think it depends who the user is. Someone who uses Rust, or someone who's familiar with Docker. I think `cargo run -- <arg1> --my-flag1` works if it's a simple binary with arguments, but for anything more complex might not work.

**Comment by jayong93** [comment]
>This is how I imagine a good way to go about implementing this:
>Setting the level, summary and body of a notification
>```
>[scripts.test-command]
>alias = 'test-command'
>command = '''
>sudo apt upgrade
>
>send-notify -u $LEVEL "$1" "$2"
>'''
>```

you don't have to change commands to pass arguments for scripts. you could just pass args via rust's `Command::arg` or `Command::args`. because when bash runs a script from text, bash pass the remain args to the script, (e.g. `bash -c "echo $0" "hello world"` will print `hello world`. passed args start with `$0`) I'm not sure how other shells work with this feature though.

So with my implementation(#56), I use pier like this:
```
[scripts.mem-usage]
alias = 'mem-usage'
command = 'ps -eo rsz,cmd | grep "^\s*[0-9]\+ $0" | grep -v grep'
```
This script just show how many memory the process I passed via `$0` uses.
```
$ pier mem-usage watch
1748 watch -n 1 echo hello world
```

**Comment by Plommonsorbet** [comment]
Not quite sure what you mean by this.
> you don't have to substitute variables in commands to pass arguments for scripts.

This was my thought as well, so glad we agree on that :)
> you could just pass args via rust's `Command::arg` or `Command::args`.

However just a slight detail, because pier creates and executes temp files for scripts that have a shebang it would be good to always start from arg `$1`. Since when executing files the `$0` is the filename it could be a bit confusing / inconsistent with different starting indices. Perhaps the `$0` when executing inline should be the `alias` or something like that.

## The cli interface
From reading through the clap docs a few times I believe that these two syntaxes are compatible. I really like the option for having the `--` syntax as it allows you to pass arguments with hyphens without having to put them in quotes.

``` shell
pier run test-alias <arg> <arg n..>
```


``` shell
pier run test-alias -- <arg> <arg n..>
```

Personally I'm not a super big fan of the `-a` syntax so my vote is just using the method described above, but if there is an argument for having both syntaxes then that could be done as well.

## Related thoughts
I think the reason I've been putting of touching this is because I want to limit the amount of breaking changes to the api/config. I do think breaking changes are fine but currently we don't have a policy for depreceating api methods/features/config options. Maybe we could write something up?

**Comment by Plommonsorbet** [comment]
What do you guys think of my proposal? @jayong93 @BenSchZA

**Comment by Plommonsorbet** [comment]
Implemented in #62
